### ISSUE TITLE
t\12-udp-open.t fails on MSWin32

### ISSUE THREAD

**Issue by leonerd** [genio/p5-UV#38]
Failure:

```
Couldn't _open (-4071): invalid argument at C:\Strawberry\cpan\build\UV-1.909-0\blib\lib/UV/UDP.pm line 17.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 25 just after 1.
t\12-udp-open.t .......... 
Dubious, test returned 25 (wstat 6400, 0x1900)
All 1 subtests passed
```

The line in the .t file is:

https://metacpan.org/release/PEVANS/UV-1.909/source/t/12-udp-open.t#L24
```perl
    $udp->open($rd);
```

which wraps the UV `uv_udp_open` function here:

https://metacpan.org/release/PEVANS/UV-1.909/source/lib/UV.xs#L2097
```C
        CHECKCALL(uv_udp_open(self->h, _MAKE_SOCK(fd)));
```

The `$rd` handle was created as a result of some `socketpair` fakery- found here:

https://metacpan.org/release/PEVANS/UV-1.909/source/t/lib/UVTestHelpers.pm#L51-75:
```perl
sub socketpair_inet_dgram
{
    my ($rd, $wr);
 
    # Maybe socketpair(2) can do it?
    ($rd, $wr) = IO::Socket->socketpair(AF_INET, SOCK_DGRAM, 0)
        and return ($rd, $wr);
 
    # If not, go the long way round
    $rd = IO::Socket::INET->new(
        LocalHost => "127.0.0.1",
        LocalPort => 0,
        Proto     => "udp",
    ) or die "Cannot socket - $@";
 
    $wr = IO::Socket::INET->new(
        PeerHost => $rd->sockhost,
        PeerPort => $rd->sockport,
        Proto    => "udp",
    ) or die "Cannot socket/connect - $@";
 
    $rd->connect($wr->sockport, inet_aton($wr->sockhost)) or die "Cannot connect - $!";
 
    return ($rd, $wr);
}
```

which apparently succeeded. It is passed via that `_MAKE_SOCK` macro which I admit I copied from the previous author while having almost no idea what it actually does:

https://metacpan.org/release/PEVANS/UV-1.909/source/lib/UV.xs#L30-35:

```C
#if defined(__MINGW32__) || defined(WIN32)
#include <io.h> /* we need _get_osfhandle() on windows */
#  define _MAKE_SOCK(f) (_get_osfhandle(f))
#else
#  define _MAKE_SOCK(f) (f)
#endif
```

**Comment by leonerd** [comment]
At this point I admit I have no idea even *if* this test is supposed to work or not. It works on every platform other than Windows, it only fails on Windows. If however it is supposed to work, I observe that it currently doesn't. And I don't know what is required to fix it to make it work.

**Comment by genio** [comment]
@xenu Any chance you can chime in on this one?

**Comment by leonerd** [comment]
(discussion raised at https://github.com/libuv/libuv/discussions/3239)

**Comment by leonerd** [comment]
@bulk88 - You've been suggested as someone who might have some ideas here also. Any thoughts?

**Comment by leonerd** [comment]
Further commentary from various places:

`uv_udp_open()` wants an argument of type `uv_os_sock_t` (http://docs.libuv.org/en/v1.x/udp.html#c.uv_udp_open), which on Windows is a `SOCKET` type. It isn't clear at all to me how we might obtain one of those from `$rd` in Perl, though it is suggested elsewhere that `_get_osfhandle()` isn't the thing because that yields a `HANDLE`. But is a `SOCKET` a `HANDLE`? I don't know...

**Comment by leonerd** [comment]
I've still no idea if we're barking up the wrong end of the tree. It may be that the `UV.xs` code is all correct and it's entirely my unit test that's wrong in the way it hacks up that `socketpair()`.

**Comment by xenu** [comment]
Good news: I have a fix which makes all tests pass (it fixes both this ticket and #35).
Bad news: It's completely on perl's side and I don't think there are any (sane) workarounds. UV must require perl 5.36 on Windows. 

https://github.com/Perl/perl5/pull/18986

**Comment by xenu** [comment]
tl;dr: libuv uses IOCP internally and IOCP requires overlapped sockets.

**Comment by leonerd** [comment]
Ahah! Well, this is indeed informational progress. I can just "skip" these two tests for now on Windows, and put a comment in to see this bug or the Perl PR for details, and maybe revisit it another time.

Thanks muchly for your efforts @xenu

**Comment by leonerd** [comment]
As of 1.910 this test is now skipped on Windows, because of this issue. Furthermore, the underlying `UV::UDP->open` method will complain about it. We can revisit if such a time comes that Perl can create WSA_FLAG_OVERLAPPED-enabled sockets.
